### DISCUSSION TITLE
How to test with a mocking framework using Entity Framework Core

### DISCUSSION THREAD

**Discussion by Kazunari001** [Avanade/Beef#259] [Q&A] [answered]
Hi team!
I am using Beef with Entity Framework Core. I want to test with a mocking framework because I want to avoid connecting to the Database when doing unit tests.
But I'm looking at [Microsoft Learning Site](https://learn.microsoft.com/en-us/ef/ef6/fundamentals/testing/mocking) and trying to adapt it to the Beef framework, but it doesn't work.
Any good ideas on this?

**Comment by chullybun** [comment, marked as answer]
Hi. Most capabilities enabled by _CoreEx_, that _Beef_ leverages, are primarily interface driven, by design, and therefore mockable. Also, the `IXxxData` is a common mocked layer in unit tests where required. 

There is no example of the EF and Database mocking as per the link you shared (and there are acknowledged [limitations](https://learn.microsoft.com/en-us/ef/ef6/fundamentals/testing/mocking#limitations-of-ef-in-memory-test-doubles) with its usage). I have a philosophical issue with the mocking of key dependencies such as its tightly coupled database, as such the preferred approach has been to unit test intra-domain and mock inter-domain (see [definition](https://github.com/Avanade/Beef/tree/master/tools/Beef.Test.NUnit#intra-domain-vs-inter-domain-testing)) dependencies where possible.

By mocking the database dependency so much of the solution is not validated. For example, existence of table(s) and/or column(s), column type mapping, identifier creation, timestamping for optimistic concurrency, duplicate checking, stored-procedure existence and related logic, data filtering and sorting validation, etc. Plus, with containers, and CI pipelines, it is relatively easy to ensure that the verification is performed early and often (see [`CI.yml`](https://github.com/Avanade/Beef/blob/master/.github/workflows/CI.yml) for example, which is used for _Beef_ itself).

In the sample, to improve the execution performance of the validation unit tests and minimize data set up requirements, the data access is mocked. See [`EmployeeValidatorTest`](https://github.com/Avanade/Beef/blob/master/samples/MyEf.Hr/MyEf.Hr.Test/Validators/EmployeeValidatorTest.cs) where this is performed.

I hope this helps :-)

**Reply by Kazunari001** [reply]
Hi.
Thank you so much for your detailed response. Thanks to you, I now have a better understanding of how to mock tests with Beef.
I'm a little worried about launching a service container on Beef's CI, but it's useful to be able to check it on the pipeline by performing operations on the DB.
(This is because GitHub Actions services currently have limitations such as being Linux-only)

I'll take a look at the sample you gave me and study it carefully.
